pkg/provisioner/fss/fss.go

```diff
@@ -84,175 +95,373 @@ func (fsp *filesystemProvisioner) getMountTargetFromID(ctx context.Context, moun
+}
+
+func (fsp *filesystemProvisioner) getFileSystemByDisplayName(ctx context.Context, ad, displayName string) (*filestorage.FileSystemSummary, error) {
+	resp, err := fsp.client.FSS().ListFileSystems(ctx, filestorage.ListFileSystemsRequest{
+		CompartmentId:      common.String(fsp.client.CompartmentOCID()),
+		AvailabilityDomain: &ad,
+		DisplayName:        &displayName,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.Items) > 1 {
+		return nil, errors.Errorf("found more than one file system with display name %q", displayName)
+	}
+
+	if len(resp.Items) == 1 {
+		return &resp.Items[0], nil
+	}
+
+	return nil, errNotFound
+}
+
+func (fsp *filesystemProvisioner) getOrCreateFileSystem(ctx context.Context, logger *zap.SugaredLogger, ad, displayName string) (*filestorage.FileSystem, error) {
+	fs, err := fsp.getFileSystemByDisplayName(ctx, ad, displayName)
+	if err != nil && err != errNotFound {
+		return nil, err
+	}
+	if fs != nil {
+		return fsp.awaitFileSystem(ctx, logger, *fs.Id)
+	}
+
+	resp, err := fsp.client.FSS().CreateFileSystem(ctx, filestorage.CreateFileSystemRequest{
+		CreateFileSystemDetails: filestorage.CreateFileSystemDetails{
 			AvailabilityDomain: &ad,
-			SubnetId:           &subnetID,
 			CompartmentId:      common.String(fsp.client.CompartmentOCID()),
-			DisplayName:        mtDisplayName,
+			DisplayName:        &displayName,
 		},
 	})
 	if err != nil {
 		return nil, err
 	}
-	return &resp.MountTarget, nil
+
+	logger.With("fileSystemID", *resp.FileSystem.Id).Info("Created filesystem")
+
+	return fsp.awaitFileSystem(ctx, logger, *resp.FileSystem.Id)
+}
+
+// findExport looks for an existing export with the same filesystem ID, export set ID, and path.
+// NOTE: No two non-'DELETED' export resources in the same export set can reference the same file system.
+func (fsp *filesystemProvisioner) findExport(ctx context.Context, fsID, exportSetID string) (*filestorage.ExportSummary, error) {
+	var page *string
+	for {
+		resp, err := fsp.client.FSS().ListExports(ctx, filestorage.ListExportsRequest{
+			CompartmentId: common.String(fsp.client.CompartmentOCID()),
+			FileSystemId:  &fsID,
+			ExportSetId:   &exportSetID,
+			Page:          page,
+		})
+		if err != nil {
+			return nil, err
+		}
+		for _, export := range resp.Items {
+			if export.LifecycleState == filestorage.ExportSummaryLifecycleStateCreating ||
+				export.LifecycleState == filestorage.ExportSummaryLifecycleStateActive {
+				return &export, nil
+			}
+		}
+		if page = resp.OpcNextPage; resp.OpcNextPage == nil {
+			break
+		}
+	}
+
+	return nil, errNotFound
+}
+
+func (fsp *filesystemProvisioner) getOrCreateExport(ctx context.Context, logger *zap.SugaredLogger, fsID, exportSetID string) (*filestorage.Export, error) {
+	export, err := fsp.findExport(ctx, fsID, exportSetID)
+	if err != nil && err != errNotFound {
+		return nil, err
+	}
+	if export != nil {
+		return fsp.awaitExport(ctx, logger, *export.Id)
+	}
+
+	path := "/" + fsID
```

Comment:
NOTE: This means that we will only set the path ourselves if there's an existing export. Perhaps we should fatally error if there's an existing export with a different path?

> No two non-'DELETED' export resources in the same export set can reference the same file system.